oxenmq/bt_producer.h

```diff
@@ -0,0 +1,306 @@
+    // key.  Does nothing in a release build.
+#ifdef NDEBUG
+    constexpr void check_incrementing_key(size_t) const {}
+#else
+    // String view into the buffer where we wrote the previous key.
+    std::string_view last_key;
+    void check_incrementing_key(size_t size);
+#endif
+
+public:
+    // Construction is identical to bt_list_producer
+    using bt_list_producer::bt_list_producer;
+
+    /// Returns a string_view into the currently serialized data buffer.  Note that the returned
+    /// view includes the `e` dict end serialization markers which will be overwritten if the dict
+    /// (or an active sublist/subdict) is appended to.
+    std::string_view view() const { return bt_list_producer::view(); }
+
+    /// Appends a key-value pair with a string or integer value.  The key must be > the last key
+    /// added, but this is only enforced (with an assertion) in debug builds.
+    template <typename T, std::enable_if_t<std::is_convertible_v<T, std::string_view> || std::is_integral_v<T>, int> = 0>
+    void append(std::string_view key, const T& value) {
+        if (has_child) throw std::logic_error{"Cannot append to list when a sublist is active"};
+        append_impl(key);
+        check_incrementing_key(key.size());
+        append_impl(value);
+        append_intermediate_ends();
+    }
+
+    /// Appends pairs from the range [from, to) to the dict.  Elements must have a .first
+    /// convertible to a string_view, and a .second that is either string view convertible or an
+    /// integer.  This does *not* append the elements as a subdict: for that you should use
+    /// something like: `l.append_dict().append(key, from, to);`
+    ///
+    /// Also note that the range *must* be sorted by keys, which means either using an ordered
+    /// container (e.g. std::map) or a manually ordered container (such as a vector or list of
+    /// pairs).  unordered_map, however, is not acceptable.
+    template <typename ForwardIt, std::enable_if_t<!std::is_convertible_v<ForwardIt, std::string_view>, int> = 0>
+    void append(ForwardIt from, ForwardIt to) {
+        if (has_child) throw std::logic_error{"Cannot append to list when a sublist is active"};
+        using KeyType = std::remove_cv_t<std::decay_t<decltype(from->first)>>;
+        using ValType = std::decay_t<decltype(from->second)>;
+        static_assert(std::is_convertible_v<decltype(from->first), std::string_view>);
+        static_assert(std::is_convertible_v<ValType, std::string_view> || std::is_integral_v<ValType>);
+        using BadUnorderedMap = std::unordered_map<KeyType, ValType>;
+        static_assert(!( // Disallow unordered_map iterators because they are not going to be ordered.
+                    std::is_same_v<typename BadUnorderedMap::iterator, ForwardIt> ||
+                    std::is_same_v<typename BadUnorderedMap::const_iterator, ForwardIt>));
+        while (from != to) {
+            auto& [k, v] = *from++;
```

Comment:
could this be a `const auto & [k, v]` instead or am i missing something?

Reply:
Just laziness. Pushed.